arrow/arrowhead.pyx

```diff
@@ -0,0 +1,149 @@
+from . cimport mersenne
+from . cimport obsidian
+
+
+np.import_array()  # Initialize numpy's C API so it won't segfault
+
+
+cdef class Arrowhead:
+    """Cython interface to the C-coded Gillespie algorithm "obsidian", leaving
+    obsidian.c free of Python dependencies.
+    """
+
+    cdef:
+        int random_seed
+        obsidian.Info info
+        tuple refs
+
+    # __cinit__() is always called immediately on construction, before CPython
+    # considers calling __init__(), which it can skip. `self` is not yet fully
+    # constructed so *BE CAREFUL* invoking Python operations that might touch it!
+    @cython.boundscheck(False)
+    @cython.wraparound(False)
+    def __cinit__(self, *args):
+        memset(&self.info, 0, sizeof(obsidian.Info))
+        self.info.random_state = <mersenne.MTState*> PyMem_Malloc(sizeof(mersenne.MTState))
+        if not self.info.random_state:
+            raise MemoryError()
+
+    # This will accept any type of array in C-contiguous layout that support
+    # the memoryview protocol (numpy array, array.array, Cython array, buffer
+    # interface) with the declared element type and number of dimensions.
+    @cython.boundscheck(False)
+    @cython.wraparound(False)
+    def __init__(
+            self,
+            int random_seed,
+            int64_t[:, ::1] stoichiometry not None,
+            double[::1] rates not None,
+
+            int64_t[::1] reactants_lengths not None,
+            int64_t[::1] reactants_indexes not None,
+            int64_t[::1] reactants not None,
+            int64_t[::1] reactions not None,
+
+            int64_t[::1] dependencies_lengths not None,
+            int64_t[::1] dependencies_indexes not None,
+            int64_t[::1] dependencies not None,
+
+            int64_t[::1] substrates_lengths not None,
+            int64_t[::1] substrates_indexes not None,
+            int64_t[::1] substrates not None,
+            *args):
+        self.random_seed = random_seed
+        mersenne.seed(self.info.random_state, random_seed)
+
+        self.info.reactions_count = stoichiometry.shape[0]
+        self.info.substrates_count = stoichiometry.shape[1]
+        self.info.stoichiometry = &stoichiometry[0, 0]
+        self.info.rates = &rates[0]
+
+        self.info.reactants_lengths = &reactants_lengths[0]
+        self.info.reactants_indexes = &reactants_indexes[0]
+        self.info.reactants = &reactants[0]
+        self.info.reactions = &reactions[0]
+
+        self.info.dependencies_lengths = &dependencies_lengths[0]
+        self.info.dependencies_indexes = &dependencies_indexes[0]
+        self.info.dependencies = &dependencies[0]
+
+        self.info.substrates_lengths = &substrates_lengths[0]
+        self.info.substrates_indexes = &substrates_indexes[0]
+        self.info.substrates = &substrates[0]
+
+        # TODO(jerry): Check array sizes? E.g. rates.shape[0] >= reactions_count
+
+        self.refs = (  # hold refs to these memoryviews while we hold their data ptrs
```

Comment:
Interesting, so if you don't hang on to these then python tries to garbage collect them? Seems like a low-level thing to need to do, did you find this out the hard way?

Reply:
Yes, needing to hold these refs is a low-level thing to do. I did not find out the hard way. What follows is my conceptual model -- if it's wrong, I sure want to hear about it!

Since Cython is a superset of Python with C added in, we have to program with the semantics of Python, and the semantics of C, and the semantics of their interactions. The C parts need C memory management. (I'll get to `__dealloc__()` below.)

When this code gets and holds raw C pointers to numpy array buffers, it has to ensure those pointers remain valid as long as needed. Python does ref counting and occasional object cycle detection. No memory compaction, so we don't have to lock any nodes from relocating. No GC tracing of pointers. Python cannot ref-count raw C pointers -- there's no place to their store ref counts and the called C code wouldn't know to inc/dec them.

Since this class has to keep those data buffers alive, the obvious approach is to hold references to the memoryview objects. The Cython compiler will handle incrementing and decrementing their ref counts.